src/ape_cache/query.py

```diff
@@ -0,0 +1,169 @@
+import math
+from functools import partial
+from pathlib import Path
+from typing import Any, Dict, Optional
+
+import pandas as pd
+from pydantic import BaseModel
+from sqlalchemy import create_engine  # type: ignore
+from sqlalchemy.sql import text  # type: ignore
+
+from ape.api import QueryAPI, QueryType
+from ape.api.query import AccountQuery, BlockQuery, ContractEventQuery, _BaseQuery
+from ape.exceptions import QueryEngineError
+from ape.utils import logger, singledispatchmethod
+
+from . import models
+
+TABLE_NAME = {
+    BlockQuery: "blocks",
+    AccountQuery: "transactions",
+    ContractEventQuery: "contract_events",
+}
+
+
+def get_columns_from_item(query: _BaseQuery, item: BaseModel) -> Dict[str, Any]:
+    return {k: v for k, v in item.dict().items() if k in query.columns}
+
+
+class CacheQueryProvider(QueryAPI):
+    """
+    Default implementation of the ape.api.query.QueryAPI
+    Allows for the query of blockchain data using connected provider
+    """
+
+    @property
+    def database_file(self) -> Path:
+        if not self.network_manager.active_provider:
+            raise QueryEngineError("Not connected to a network")
+
+        ecosystem_name = self.network_manager.active_provider.network.ecosystem.name
+        network_name = self.network_manager.active_provider.network.name
+        if network_name == "local":
+            raise QueryEngineError("Cannot cache local network")
+
+        if "-fork" in network_name:
+            network_name = network_name.replace("-fork", "")
+
+        if not (self.config_manager.DATA_FOLDER / ecosystem_name).exists():
+            (self.config_manager.DATA_FOLDER / ecosystem_name).mkdir()
+
+        return self.config_manager.DATA_FOLDER / ecosystem_name / f"{network_name}.db"
+
+    @property
+    def engine(self):
+        return create_engine(f"sqlite:///{self.database_file}", pool_pre_ping=True)
+
+    def init_db(self):
+        if self.database_file.is_file():
+            raise QueryEngineError("Database has already been initialized")
+
+        models.Base.metadata.create_all(bind=self.engine)
+
+    def purge_db(self):
+        if not self.database_file.is_file():
+            # Add check here to show we have a file that exists
+            raise QueryEngineError("Database must be initialized")
+
+        self.database_file.unlink()
+
+    @singledispatchmethod
+    def estimate_query(self, query: QueryType) -> Optional[int]:  # type: ignore
+        return None  # can't handle this query
+
+    @estimate_query.register
+    def estimate_block_query(self, query: BlockQuery) -> Optional[int]:
+        try:
```

Comment:
`try..except` statements are important here to ensure there is nothing that stops "normal operation" of brute force querying the providers